File: 3. Longest Substring Without Repeating/longest_substring_without_repeating_substring.rb
s1 = "abcabcbb"
s2 = "bbbbb"
s3 = "pwwkew"
s4 = " ad"

def length_of_longest_substring(s)
  streak = 0
  hash = Hash.new

  s.each_char.with_index do |char, i|
    hash.reject! &-> k,v { v < hash[char] } if hash.include?(char)
    hash[char] = i
    streak = hash.size if hash.size > streak
  end
  
  streak
end

pp length_of_longest_substring s1
pp length_of_longest_substring s2
pp length_of_longest_substring s3
pp length_of_longest_substring s4

File: 3. Longest Substring Without Repeating/longest_substring_without_repeating_substring.ts
function lengthOfLongestSubstring(s: string): number {
  let streak: number = 0;
  let i = 0;
  let map = new Map<string, number>();

  for (let char of s) {
    if (map.has(char)) {
      let char_i: number | undefined = map.get(char);
      if (char_i) {
        for (let [k, v] of map.entries()) {
          if (v < char_i) map.delete(k);
        }
      }
    }

    map.set(char, i);
    if (streak < map.size) streak = map.size;
    i++;
  }

  return streak;
};

const s1: string = "abcabcbb";
const s2: string = "bbbbb";
const s3: string = "pwwkew";
const s4: string = "";
const s5: string = "dvdf";

console.log(lengthOfLongestSubstring(s1));
console.log(lengthOfLongestSubstring(s2));
console.log(lengthOfLongestSubstring(s3));
console.log(lengthOfLongestSubstring(s4));
console.log(lengthOfLongestSubstring(s5));
